mautic/mautic
# Support tokens in an email entity's "from address" override

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR adds the feature to use a custom field value as the email's "from address" on the advanced tab. This also does a little refactoring in order to use a testable class. 

Edit: I (@escopecz ) have updated this PR to work with the current code as it was written 5 years ago and it required some refactoring and added more tests.

[//]: # ( As applicable: )

#### Steps to test this PR:
1. Create a new custom email field.
2. Create a new email and on the advanced tab, test that the form address will validate with an email address or a custom field token `{contactfield=other_email}` but will reject anything that is not
3. Edit a contact and fill in an email address different than the system's default or the user's email if mail is owner is enabled. 
4. Create a segment with some contacts and send the email
5. The emails should arrive from the appropriate email addresses, if they had a token value, then it should be from that email, if the contact has an owner and mail is owner is enabled then from the owner, otherwise from the system default. 

Test with the following scenarios with both a SMTP mailer _and_ a token mailer such as Sparkpost or Sendgrid to ensure that the hierarchy below is obeyed. 
1. With a token as the advanced tab's from email and mailer is owner enabled.
2. With a token as the advanced tab's from email and mailer is owner disabled.
3. With a normal email address as the advanced tab's from email and mailer is owner enabled
4. With a normal email address as the advanced tab's from email and mailer is owner disabled
5. Nothing in the advanced tab's from email and mailer is owner enabled
6. Nothing in the advanced tab's from email and mailer is owner disabled

The hierarchy is as follows:
1. Email on advanced tab with custom field token if the value exists (otherwise it proceeds to the owner if applicable or default if no owner/owner is disabled)
2. Owner if mail is owner is enabled and contact is assigned to owner
2. Email on advanced tab with a standard email address defined
4. System default
